helix-core/src/main/java/org/apache/helix/model/TrieClusterTopology.java

```diff
@@ -0,0 +1,244 @@
+   * @param node the node from where to collect all the nodes' paths.
+   * @return All the paths under the node.
+   */
+  private Set<String> getPathUnderNode(TrieNode node) {
+    Set<String> resultMap = new HashSet<>();
+    Deque<TrieNode> nodeStack = new ArrayDeque<>();
+    nodeStack.push(node);
+    while (!nodeStack.isEmpty()) {
+      node = nodeStack.pop();
+      if (node.getChildren().isEmpty()) {
+        resultMap.add(node.getPath());
+      } else {
+        for (TrieNode child : node.getChildren().values()) {
+          nodeStack.push(child);
+        }
+      }
+    }
+    return resultMap;
+  }
+
+  private TrieNode getStartNode(LinkedHashMap<String, String> domain) {
+    TrieNode curNode = _rootNode;
+    TrieNode nextNode;
+    for (Map.Entry<String, String> entry : domain.entrySet()) {
+      nextNode = curNode.getChildren().get(entry.getKey() + CONNECTOR + entry.getValue());
+      if (nextNode == null) {
+        throw new IllegalArgumentException(String
+            .format("The input domain %s does not have the value %s", entry.getKey(),
+                entry.getValue()));
+      }
+      curNode = nextNode;
+    }
+    return curNode;
+  }
+
+  private void removeInvalidInstanceConfig(Map<String, InstanceConfig> instanceConfigMap) {
+    Set<String> toRemoveConfig = new HashSet<>();
+    for (String instanceName : instanceConfigMap.keySet()) {
+      Map<String, String> domainAsMap = instanceConfigMap.get(instanceName).getDomainAsMap();
+      if (domainAsMap.isEmpty()) {
+        logger.info(String.format("Domain for instance %s is not set", instanceName));
+        toRemoveConfig.add(instanceName);
+      } else {
+        for (String key : _topologyKeys) {
+          String value = domainAsMap.get(key);
+          if (value == null || value.length() == 0) {
+            logger.info(String.format("Domain %s for instance %s is not set",
+                domainAsMap.get(key), instanceName));
+            toRemoveConfig.add(instanceName);
+            break;
+          }
+        }
+      }
+    }
+    toRemoveConfig.forEach(entry -> instanceConfigMap.remove(entry));
+  }
+
+  /**
+   * Constructs a trie based on the provided instance config map. It loops through all instance
+   * configs and constructs the trie in a top down manner.
+   */
+  private void constructTrie(Map<String, InstanceConfig> instanceConfigMap) {
+    removeInvalidInstanceConfig(instanceConfigMap);
+    Map<String, Map<String, String>> instanceDomainsMap = new HashMap<>();
+    instanceConfigMap.entrySet().forEach(
+        entry -> instanceDomainsMap.put(entry.getKey(), entry.getValue().getDomainAsMap()));
+
+    for (Map.Entry<String, Map<String, String>> entry : instanceDomainsMap.entrySet()) {
+      TrieNode curNode = _rootNode;
+      String path = "";
+      for (int i = 0; i < _topologyKeys.length; i++) {
+        String key = _topologyKeys[i] + CONNECTOR + entry.getValue().get(_topologyKeys[i]);
```

Comment:
Thanks for the detailed reply.

I think the previous 3 points can be achieved by adding a flag and using non static 'DEFAULT_DOMAIN_PREFIX'. 
For the point 4, 
Please correct me if I am wrong, line 288 in constructTrie will also add a "default" node similar to 'clusterTopologyKeyDefaultValue'. Maybe we could read the tree and bypass default node. 


In my own opinion, I think the question would be, do we construct topology tree in class Topology for these instances? If yes, its better to read from the existing tree. If not, I think it's ok to have another orthogonal data structure representing the hierarchy.

Reply:
There're some offline discussion, and the conclusion is that trie is a good fit for cluster topology, as also mentioned here: https://github.com/apache/helix/issues/447
and should apply to both rebalance and user facing APIs, meaning that we only need one class for both uses. 
In this PR, we will create a generic trie structure with validation and construction logic. Topology APIs will use this structure and do whatever traverse needed. Laster, `topology` logic will also use this trie structure and implement the functions it need. The new trie structure is not fundamentally different from the tree in `topology` now, but with a clearer organization/validation logic and present in a more generic way. `topology` should be able to use it, if something mismatch, we can always come back to adjust the structure.